petab/core.py

```diff
@@ -321,25 +438,52 @@ def to_float_if_float(x):
 
 
 def get_num_placeholders(formula_string, observable_id, override_type):
-    """Get number of unique placeholder variables in noise or observable definition"""
-
+    """
+    Get number of unique placeholder variables in noise or observable
+    definition.
+    """
+    
     pattern = re.compile(re.escape(override_type) + r'Parameter\d+_' + re.escape(observable_id))
     placeholders = set()
     for free_sym in sp.sympify(formula_string).free_symbols:
         free_sym = str(free_sym)
         if pattern.match(free_sym):
             placeholders.add(free_sym)
+    
     return len(placeholders)
 
 
 def get_dynamic_parameters_from_sbml(sbml_model):
-    """Get list of non-constant parameters in sbml model"""
-
+    """
+    Get list of non-constant parameters in sbml model.
+    TODO: This is not what Y would expect. Remove?
+    """
     return [p.getId() for p in sbml_model.getListOfParameters() if not p.getConstant()]
 
 
-def get_notnull_columns(df, candidates):
-    """"Return list of df-columns in candidates which are not all null/nan
+def get_dynamic_simulation_parameters(sbml_model, parameter_df):
+    par_opt_ids = get_optimization_parameters(parameter_df)
+    sbml_pars = sbml_model.getListOfParameters()
+    par_ids = [
+        p.getId() for p in sbml_pars
+        if p.getId() in par_opt_ids
```

Comment:
Please reverse condition order. comparing a single string is cheaper than lookup in a potentially long list